### DISCUSSION TITLE
FS.writeFile is writing incorrect data when passed a string containing binary data.

### DISCUSSION THREAD

**Discussion by 98ahni** [emscripten-core/emscripten#22056] [Q&A] [answered]
I have a string with the contents of a binary file, but when I write this string to a file using `FS.writeFile("/WorkDir/" + file.name, res.body);` the file gets corrupted for some reason. The file should be 7,203,723 bytes long, instead it becomes 10,480,395 bytes once saved. The extra bytes are seemingly inserted randomly through the file making it unreadable.

**Comment by sbc100** [comment]
I think this is because writeFile is doing to first UTF8 encode the string and then write the result, which is not what you want for a binary file.

Assuming `res` is the result of a fetch request I imagine you probably await the `res.arrayBuffer()`  and pass that to `FS.writeFile`.

**Reply by sbc100** [reply]
See https://github.com/emscripten-core/emscripten/blob/3de1b9091adbadc46706959bd97efa033214bc2c/src/library_fs.js#L1310-L1318

**Reply by 98ahni** [reply]
The response `res` is a json object, which doesn't have the `arrayBuffer()` function. If I run `JSON.stringify(res);` I get:  
``` JSON
{
    "result":false,
    "body":"ID3\u0004\u0000\u0000\u0000\u0000\u0001;TXXX[...]",
    "headers":{
        "cache-control":"private, max-age=0, must-revalidate",
        "content-encoding":"gzip",
        "content-type":"text/plain;charset=UTF-8",
        "date":"Wed, 05 Jun 2024 07:50:29 GMT",
        "expires":"Wed, 05 Jun 2024 07:50:29 GMT",
        "server":"ESF",
        "vary":"Origin, X-Origin",
        "x-guploader-uploadid":"[...]",
        "Content-Type":"audio/mpeg"
    },
    "status":200,
    "statusText":null
}
```
The exact same code works for a *.txt file and it has the same response as above, just a different `"body"` and `"Content-Type"`. The response even says that it uses UTF-8.

**Reply by sbc100** [reply]
What does Object.keys(res) show?  Are you sure it doesn't have an arrayBuffer() method?  Where does res come from?  It kind of looks like the result of fetch?  

In any case, what you need here is an ArrayBuffer representing the binary, no the JS string form.

**Reply by 98ahni** [reply]
`Object.keys(res)` returns `(5) ['result', 'body', 'headers', 'status', 'statusText']`.

The response comes from:  
```JS
const res = await gapi.client.drive.files.get({
    'fileId': file.id,
    'alt': 'media'
});
```
It always comes as a string.

**Reply by sbc100** [reply]
You should be able to convert that string to an ArrayBuffer and then pass that to `FS.writeFile`

**Reply by 98ahni** [reply]
How do I do that?

**Comment by 98ahni** [comment, marked as answer]
This seems to be one of many oversights from Google and the documentation they provide. The `"body"` is a File blob according to the docs. It does not specify _how_ to choose whether it's represented as a string or an ArrayBuffer, just that you can. Apparently, this is set in the `Google.drive()` constructor which **doesn't exist** in the JS version of the api. The discussions I could find for fixing this ended with one dev team blaming another.  
Sorry for the mini rant, on to my "solution".  
  
I managed to brute-force it by looping over the string, getting each char and then passing that to a Uint8Array. 
```JS
const res = await gapi.client.drive.files.get({
    'fileId': file.id,
    'alt': 'media'
});
var bytes = [];
for (var i = 0; i < res.body.length; ++i) {
    bytes.push(res.body.charCodeAt(i));
}
FS.writeFile("/GoogleDrive/" + file.name, new Uint8Array(bytes));
```
Which works, but it feels like the wrong solution. I'm open to suggestions on a better way.

**Reply by sbc100** [reply]
That looks like the correct solution to me.
